File: src/level.cpp
#include "level.hpp"
#include "settings.hpp"
#include "support.hpp"
#include "tile.hpp"
#include "enemy.hpp"
#include "weapon.hpp"
#include <SFML/System/Vector2.hpp>
#include <algorithm>
#include <cassert>
#include <cstdlib>
#include <iterator>
#include <map>
#include <string>
#include <utility>
#include <vector>
#include <memory>
#include <iostream>

Level::Level()
    : game_paused(false),
      animation_player(std::make_shared<AnimationPlayer>(*this)),
      magic_player(std::make_shared<MagicPlayer>(this->animation_player))
{
  this->create_map();
}

void Level::create_map()
{
  std::map<std::string, std::vector<std::vector<std::string>>> layouts = {
      {"boundary", import_csv_layout("./src/map/map_FloorBlocks.csv")},
      {"grass", import_csv_layout("./src/map/map_Grass.csv")},
      {"object", import_csv_layout("./src/map/map_Objects.csv")},
      {"entities", import_csv_layout("./src/map/map_Entities.csv")},
  };

  std::map<std::string, std::vector<std::shared_ptr<SpriteTexture>>> graphics = {
      {"grass", import_folder("./src/graphics/grass/")},
      {"object", import_folder("./src/graphics/objects/")}};

  for (const auto &entry : layouts)
  {
    auto style = entry.first;
    auto layout = entry.second;
    int row_index = 0;
    for (const auto &row : layout)
    {
      int col_index = 0;
      for (const auto &col : row)
      {
        if (!col.compare("-1"))
        {
          col_index++;
          continue;
        }

        auto x = col_index * TILESIZE;
        auto y = row_index * TILESIZE;

        if (!style.compare("boundary"))
        {
          auto tile = std::make_shared<Tile>(sf::Vector2f(x, y), "invisible");
          this->obstacle_sprites.push_back(tile);
        }
        else if (!style.compare("grass"))
        {
          auto grass = this->random_choice<std::vector<std::shared_ptr<SpriteTexture>>>(graphics["grass"]);
          auto tile = std::make_shared<Tile>(sf::Vector2f(x, y), "grass", *grass);
          this->visible_sprites.push_back(tile);
          this->obstacle_sprites.push_back(tile);
          this->attackable_sprites.push_back(tile);
        }
        else if (!style.compare("object"))
        {
          auto sprite = graphics["object"][std::stoi(col)];
          auto tile = std::make_shared<Tile>(sf::Vector2f(x, y), "object", *sprite);
          this->visible_sprites.push_back(tile);
          this->obstacle_sprites.push_back(tile);
        }
        else if (!style.compare("entities"))
        {
          if (!col.compare("394"))
          {
            this->player = std::make_shared<Player>(
                sf::Vector2f(x, y),
                this->obstacle_sprites,
                [this]() -> void
                { this->create_attack(); },
                [this]() -> void
                {
                  this->destroy_attack();
                },
                [this](const std::string &style, int strength, int cost)
                {
                  this->create_magic(style, strength, cost);
                },
                [this]() -> void
                {
                  this->destroy_magic();
                },
                [this](int amount, const std::string &attack_type) -> void
                { this->damage_player(amount, attack_type); });
            this->upgrade = std::make_shared<Upgrade>(this->player);
            this->visible_sprites.push_back(this->player);
          }
        }

        col_index++;
      }
      row_index++;
    }
  }

  for (const auto &entry : layouts)
  {
    auto style = entry.first;
    auto layout = entry.second;
    int row_index = 0;
    for (const auto &row : layout)
    {
      int col_index = 0;
      for (const auto &col : row)
      {
        if (!col.compare("-1"))
        {
          col_index++;
          continue;
        }

        auto x = col_index * TILESIZE;
        auto y = row_index * TILESIZE;

        if (!style.compare("entities"))
        {
          if (col.compare("394")) // not 394
          {
            auto monster_name = [&col]() -> std::string
            {
              if (!col.compare("390"))
                return "bamboo";
              if (!col.compare("391"))
                return "spirit";
              if (!col.compare("392"))
                return "raccoon";
              if (!col.compare("393"))
                return "squid";
              assert(false);
            }();
            if (!monster_name.empty())
            {
              auto enemy = std::make_shared<Enemy>(
                  monster_name,
                  sf::Vector2f(x, y),
                  this->obstacle_sprites,
                  [this](const sf::Vector2f &pos, const std::string &monster_name)
                  {
                    this->trigger_death_particles(pos, monster_name);
                  },
                  [this](int exp)
                  {
                    this->add_exp(exp);
                  },
                  *this,
                  this->player);
              this->visible_sprites.push_back(enemy);
              this->attackable_sprites.push_back(enemy);
            }
          }
        }
        
        col_index++;
      }
      row_index++;
    }
  }
}

void Level::trigger_death_particles(const sf::Vector2f &pos, const std::string &particle_type)
{
  auto sprite = this->animation_player->create_particles(pos, particle_type);
  this->visible_sprites.push_back(sprite);
}

void Level::add_exp(int amount)
{
  this->player->exp(this->player->exp() + amount);
}

void Level::create_attack()
{
  this->current_attack = std::make_shared<Weapon>(this->player);
  this->visible_sprites.push_back(this->current_attack);
  this->attack_sprites.push_back(this->current_attack);
}

void Level::destroy_attack()
{
  if (this->current_attack)
  {
    this->kill(this->current_attack);
    this->current_attack = nullptr;
  }
}

void Level::create_magic(const std::string &style, int strength, int cost)
{
  if (!style.compare("heal"))
  {
    auto sprites = this->magic_player->heal(this->player, strength, cost);
    for (const auto &sprite : sprites)
    {
      this->visible_sprites.push_back(sprite);
    }
  }
  else if (!style.compare("flame"))
  {
    auto sprites = this->magic_player->flame(this->player, cost);
    for (const auto &sprite : sprites)
    {
      this->visible_sprites.push_back(sprite);
      this->attack_sprites.push_back(sprite);
    }
  }
}

void Level::destroy_magic()
{
}

void Level::run()
{
  this->visible_sprites.custom_draw(player);
  this->ui.display(this->player);

  if (this->game_paused)
  {
    this->upgrade->display();
  }
  else
  {
    this->visible_sprites.update();
    this->player_attack_logic();
    this->remove_dead_sprites();
  }
}

void Level::player_attack_logic()
{
  for (const auto &attack_sprite : this->attack_sprites)
  {
    auto collision_sprites = ::spritecollide(attack_sprite, this->attackable_sprites);
    for (const auto &collision_sprite : collision_sprites)
    {
      if (!collision_sprite->sprite_type().compare("grass"))
      {
        auto rect = collision_sprite->rect();
        auto pos = rect.center();
        auto offset = py::Vector2f(0, 75);
        int range = 3 + (rand() % 4); // 3 ~ 6
        for (auto i = 0; i < range; ++i)
        {
          auto particle = this->animation_player->create_grass_particles(pos - offset);
          this->visible_sprites.push_back(particle);
        }

        this->kill(collision_sprite);
      }
      else if (!collision_sprite->sprite_type().compare("enemy"))
      {
        auto enemy = std::static_pointer_cast<Enemy>(collision_sprite);
        enemy->get_damage(this->player, attack_sprite->sprite_type());
      }
    }
  }
}

void Level::damage_player(int amount, const std::string &attack_type)
{
  if (this->player->vulernable())
  {
    this->player->health(this->player->health() - amount);
    this->player->vulernable(false);
    this->player->restart_hurt_time();
    auto rect = this->player->rect();
    auto pos = rect.center();
    auto sprite = this->animation_player->create_particles(pos, attack_type);
    this->visible_sprites.push_back(sprite);
  }
}

YSortCameraGroup::YSortCameraGroup()
{
  auto &rnderer = GameWindow::instance();
  this->half_width = rnderer.screen().getSize().x / 2;
  this->half_height = rnderer.screen().getSize().y / 2;
  this->offset = sf::Vector2f(0, 0);

  // creating the floor
  auto r = this->floor.loadFromFile("./src/graphics/tilemap/ground.png");
  assert(r);
}

void YSortCameraGroup::custom_draw(std::shared_ptr<Player> player)
{
  this->offset.x = player->rect().centerx - this->half_width;
  this->offset.y = player->rect().centery - this->half_height;

  // drawing the floor
  auto floor_offset_pos = sf::Vector2f(this->floor.rect().left - this->offset.x, this->floor.rect().top - this->offset.y);
  this->floor.set_origin(floor_offset_pos);
  GameWindow::instance().screen().draw(this->floor.surf());

  this->sprites.sort(
      [](auto a, auto b) -> bool
      {
        return a->rect().centery < b->rect().centery;
      });

  for (const auto &sprite : this->sprites)
  {
    auto offset_rect = sf::Vector2f(sprite->rect().left - this->offset.x, sprite->rect().top - this->offset.y);
    sprite->set_origin(offset_rect);
    GameWindow::instance().screen().draw(sprite->surf());

#if 0 // draw hitbox (for debugging)
    sf::RectangleShape rectangle;
    rectangle.setSize(sf::Vector2f(sprite->rect().getSize()));
    rectangle.setOutlineColor(sf::Color::Blue);
    rectangle.setOutlineThickness(5);
    rectangle.setFillColor(sf::Color::Transparent);
    rectangle.setPosition(sf::Vector2f(sprite->rect().getPosition()));
    rectangle.setOrigin(-floor_offset_pos);

    sf::RectangleShape hitbox;
    hitbox.setSize(sf::Vector2f(sprite->hitbox().getSize()));
    hitbox.setOutlineColor(sf::Color::Red);
    hitbox.setOutlineThickness(3);
    hitbox.setFillColor(sf::Color::Transparent);
    hitbox.setPosition(sf::Vector2f(sprite->hitbox().getPosition()));
    hitbox.setOrigin(-floor_offset_pos);

    GameWindow::instance().screen().draw(rectangle);
    GameWindow::instance().screen().draw(hitbox);
#endif
  }

  GameWindow::instance().screen().display();
}

void YSortCameraGroup::update()
{
  for (const auto &sprite : this->sprites)
  {
    sprite->update();
  }
}

File: pyZelda/level.py
import enum
import random
import pygame
from enemy import Enemy
from particles import AnimationPlayer
from player import Player
from debug import debug

from settings import TILESIZE
from tile import Tile
from support import *
from ui import UI
from weapon import Weapon

from magic import MagicPlayer
from upgrade import Upgrade

class Level:
    def __init__(self) -> None:
        self.game_paused = False

        self.display_surface = pygame.display.get_surface()
        self.visible_sprites = YSortCameraGroup()
        self.obstacles_sprites = pygame.sprite.Group()
        self.attack_sprites = pygame.sprite.Group()
        self.attackable_sprites = pygame.sprite.Group()

        self.current_attck = None

        self.create_map()

        self.ui = UI()
        self.upgrade = Upgrade(self.player)

        # particles
        self.animation_player = AnimationPlayer()

        self.magic_player = MagicPlayer(self.animation_player)

    def create_map(self) -> None:
        layouts = {
            'boundary': import_csv_layout("./map/map_FloorBlocks.csv"),
            'grass': import_csv_layout("./map/map_Grass.csv"),
            'object': import_csv_layout("./map/map_Objects.csv"),
            'entities': import_csv_layout("./map/map_Entities.csv"),
        }
        graphics = {
            'grass': import_folder('./graphics/grass/'),
            'object': import_folder('./graphics/objects/'),
        }

        for style, layout in layouts.items():
            for row_index, row in enumerate(layout):
                for col_index, col in enumerate(row):
                    if col == '-1': continue
                    x = col_index * TILESIZE
                    y = row_index * TILESIZE

                    if style == 'boundary':
                        Tile((x, y), [self.obstacles_sprites], 'invisible')
                    if style == 'grass':
                        grass = random.choice(graphics['grass'])
                        Tile((x, y), [self.visible_sprites, self.obstacles_sprites, self.attackable_sprites], 'grass', grass)
                    if style == 'object':
                        surf = graphics['object'][int(col)]
                        Tile((x, y), [self.visible_sprites, self.obstacles_sprites], 'object', surf)
                    if style == 'entities':
                        if col == '394':
                            self.player = Player((x, y), 
                                [self.visible_sprites], self.obstacles_sprites, 
                                self.create_attack, self.destroy_attack,
                                self.create_magic, self.destroy_magic,
                                self.damage_player)
                        else:
                            if col == '390': monster_name = 'bamboo'
                            if col == '391': monster_name = 'spirit'
                            if col == '392': monster_name = 'raccoon'
                            if col == '393': monster_name = 'squid'
                            Enemy(monster_name, (x, y), 
                                [self.visible_sprites, self.attackable_sprites], 
                                self.obstacles_sprites, 
                                self.trigger_death_particles, 
                                self.add_xp)

        #         if col == "x":
        #             Tile((x, y), [self.visible_sprites, self.obstacles_sprites])
        #         if col == 'p':
        #             self.player = Player((x, y), [self.visible_sprites], self.obstacles_sprites)

    def toggle_menu(self):
        self.game_paused = not self.game_paused

    def create_attack(self):
        self.current_attck = Weapon(self.player, [self.visible_sprites, self.attack_sprites])

    def damage_player(self, amount, attack_type):
        if self.player.vulernable:
            self.player.health -= amount
            self.player.vulernable = False
            self.player.hurt_time = pygame.time.get_ticks()
            pos = self.player.rect.center
            self.animation_player.create_particles(attack_type, pos, [self.visible_sprites])

    def trigger_death_particles(self, pos, particle_type):
        self.animation_player.create_particles(particle_type, pos, [self.visible_sprites])

    def destroy_attack(self):
        if self.current_attck:
            self.current_attck.kill()
            self.current_attck = None

    def add_xp(self, amount):
        self.player.exp += amount

    def create_magic(self, style, strength, cost):
        # self.current_attck = Weapon(self.player, [self.visible_sprites])

        if style == 'heal':
            self.magic_player.heal(self.player, strength, cost, [self.visible_sprites])
            pass

        if style == 'flame':
            self.magic_player.flame(self.player, cost, [self.visible_sprites, self.attack_sprites])
            pass

        # print(style)
        # print(strength)
        # print(cost)

    def destroy_magic(self):
        pass
        # if self.current_attck:
        #     self.current_attck.kill()
        #     self.current_attck = None

    def player_attack_logic(self):
        if self.attack_sprites:
            for attack_sprite in self.attack_sprites:
                collision_sprites = pygame.sprite.spritecollide(attack_sprite, self.attackable_sprites, False)
                if collision_sprites:
                    for collision_sprite in collision_sprites:
                        if collision_sprite.sprite_type == 'grass':
                            pos = collision_sprite.rect.center
                            offset = pygame.math.Vector2(0, 75)
                            for leaf in range(random.randint(3, 6)):
                                self.animation_player.create_grass_particles(pos - offset, [self.visible_sprites])
                            collision_sprite.kill()
                        elif collision_sprite.sprite_type == 'enemy':
                            collision_sprite.get_damage(self.player, attack_sprite.sprite_type)


    def run(self):
        self.visible_sprites.custom_draw(self.player)
        self.ui.display(self.player)

        if self.game_paused:
            self.upgrade.display()
        else:
            self.visible_sprites.update()  # call Player's update()
            self.visible_sprites.enemy_update(self.player)
            self.player_attack_logic()

class YSortCameraGroup(pygame.sprite.Group):
    def __init__(self) -> None:
        super().__init__()
        self.display_surface = pygame.display.get_surface()
        self.half_width = self.display_surface.get_size()[0] // 2
        self.half_height = self.display_surface.get_size()[1] // 2
        self.offset = pygame.math.Vector2(0, 0)

        # creating the floor
        self.floor_surf = pygame.image.load('./graphics/tilemap/ground.png').convert()
        self.floor_rect = self.floor_surf.get_rect(topleft=(0, 0))

    def custom_draw(self, player):
        self.offset.x = player.rect.centerx - self.half_width
        self.offset.y = player.rect.centery - self.half_height

        # drawing the floor
        floor_offset_pos = self.floor_rect.topleft - self.offset
        self.display_surface.blit(self.floor_surf, floor_offset_pos)

        for sprite in sorted(self.sprites(), key=lambda sprite: sprite.rect.centery):
            offset_rect = sprite.rect.topleft - self.offset
            self.display_surface.blit(sprite.image, offset_rect)

    def enemy_update(self, player: Player):
        enemy_sprites = [sprite for sprite in self.sprites() if hasattr(sprite, 'sprite_type') and sprite.sprite_type == 'enemy']
        for sprite in enemy_sprites:
            sprite.enemy_update(player)
